---
title: "R Notebook"
#output: html_notebook
---

# Salmon project

By: Amalia de Juana, Louise Stauber Näslund, Hanna Linn

This is a recreation of parts of the paper:
Jin Y, Olsen RE, Harvey TN, et al.
Comparative transcriptomics reveals domestication-associated
features of Atlantic salmon lipid metabolism. Mol Ecol.
2020;29:1860–1872. https://doi.org/10.1111/mec.15446

"As a team, you will write a short report (max 1500-2000 words, try to be concise!) in article format (Introduction, results, discussion, methods sections at least should be there)"

Setting up an environment:
```{bash}
export PATH="/home/student14/miniconda3/bin:$PATH"
conda create --name salmonenvironment
source activate salmonenvironment
conda install -c bioconda htseq cutadapt star gffread
```
Also source ~/.bashrc if above does not work.


Restarting the environment:
```{bash}
source ~/.bashrc
source activate salmonenvironment
```


## Introduction

Domestication is a thing that happens
Explain different type of diets
Our aim:
Investigate the effects of domestication on metabolism in Atlantic salmon.
Explore the hypothesis that the change to a vegetable oil diet has unintentionally selected for a domestication-specific lipid metabolism.


## Methods


### Data analysis

```{bash}
# directories
mkdir data
mkdir data/salmonGenomeIndex

mkdir data/salmonFastqsDomesticated
mkdir data/salmonFastqsWild

mkdir data/salmonTrimmedFastqsDomesticated
mkdir data/salmonTrimmedFastqsWild

mkdir data/salmonCountsDomesticated
mkdir data/salmonCountsWild
```

0. Get data Domesticated
Filter for:
* diet: fish oil
* tissue: pyloric caeca
* strain: domesticated

```{bash}
cat E-MTAB-8306.sdrf.txt | grep -P "\tdomesticated" | grep -P "\tfish\soil" | grep -P "\tpyloric\scaecum" | cut -f37 | tail -n +2 > data/salmonFastqsDomesticated/domesticated.txt
cd data/salmonFastqsDomesticated/
wget -N -i domesticated.txt
```

0. Get data Wild
Filter for:
* diet: fish oil
* tissue: pyloric caeca
* strain: wild

(running)
```{bash} 
cat E-MTAB-8306.sdrf.txt | grep -P "\twild" | grep -P "\tfish\soil" | grep -P "\tpyloric\scaecum" | cut -f37 | tail -n +2 > data/salmonFastqsWild/wild.txt
cd data/salmonFastqsWild/
wget -N -i wild.txt
```



Preprocess data

In bash:
1. REMOVE ADAPTERS AND FILTER DATA (method ok, data usure, no space)
Get FASTQ-files from https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-8306/.
Remove adapters in the FASTQ-files with cutadapt. Using Read 1 from https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/experiment-design/illumina-adapter-sequences-1000000002694-14.pdf
AGATCGGAAGAGCACACGTCTGAACTCCAGTCA. Here it also filter out read shorter than 40 nucleotides and reads with quality score below 20.

Domesticated:
```{bash}
for i in {3522994..3522997} {3523017..3523025} {3523030..3523033};
do cutadapt --report=minimal -q 20 -m 40 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA data/salmonFastqsDomesticated/ERR${i}.fastq.gz > data/salmonTrimmedFastqsDomesticated/ERR${i}_trimmed.fastq.gz;
done

#```
#Wild:
#```{bash}
for i in {3523032..3523037};
do cutadapt --report=minimal -q 20 -m 40 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA data/salmonFastqsWild/ERR${i}.fastq.gz > data/salmonTrimmedFastqsWild/ERR${i}_trimmed.fastq.gz;
done


```


2. ALIGN
Get the FASTA-files with the Salmo Salar reference genome (ICSASG_v2) from https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/233/375/GCF_000233375.1_ICSASG_v2/GCF_000233375.1_ICSASG_v2_genomic.fna.gz

```{bash}
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/233/375/GCF_000233375.1_ICSASG_v2/GCF_000233375.1_ICSASG_v2_genomic.fna.gz
mkdir data/salmonReferenceGenome
mv GCF_000233375.1_ICSASG_v2_genomic.fna.gz data/salmonReferenceGenome

```

Get the GFF-file for the genome annotation https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/233/375/GCF_000233375.1_ICSASG_v2/GCF_000233375.1_ICSASG_v2_genomic.gff.gz

```{bash}
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/233/375/GCF_000233375.1_ICSASG_v2/GCF_000233375.1_ICSASG_v2_genomic.gff.gz
mkdir data/salmonAnnotationGenome
mv GCF_000233375.1_ICSASG_v2_genomic.gff.gz data/salmonAnnotationGenome

```

Annotate the reference genome with star.

```{bash}
# GFF to GTF, unziped
gffread <(gunzip -c data/salmonAnnotationGenome/GCF_000233375.1_ICSASG_v2_genomic.gff.gz) -T -o data/salmonAnnotationGenome/GCF_000233375.1_ICSASG_v2_genomic.gtf

# unzip the FASTA-file
gunzip data/salmonReferenceGenome/GCF_000233375.1_ICSASG_v2_genomic.fna.gz

# fixing memory error
grep ">" data/salmonReferenceGenome/GCF_000233375.1_ICSASG_v2_genomic.fna | wc -l # gives 232155
# set --genomeChrBinNbits = min(18,log2[max(GenomeLength/NumberOfRef, ReadLength)])
# NumberOfRef = 232155
# GenomeLength = 2966890203 (from webpage where the ref genome is from)
# ReadLength = 100 (from paper)
# gives --genomeChrBinNbits = 14


# create a genome annotation index
STAR --runMode genomeGenerate --genomeDir data/salmonGenomeIndex/ --genomeFastaFiles data/salmonReferenceGenome/GCF_000233375.1_ICSASG_v2_genomic.fna --sjdbGTFfile data/salmonAnnotationGenome/GCF_000233375.1_ICSASG_v2_genomic.gtf --sjdbOverhang 99 --genomeChrBinNbits 14 --genomeSAsparseD 2 --limitGenomeGenerateRAM 20000000000 --runThreadN 8

```

Align all read sequences from FASTQ-files to FASTA-files with star.

```{bash}
# (not run)
# test one first!!!!


# move ERR${i}_trimmed.fastq.gz to /data/salmonTrimmedFastqs and delete 
# map the trimmed sequences to the index

# to reduce computations, don't use:
# Domesticated: {19...25} {3523030..3523033} 
# Wild: {3522996..3522997} {17..31}

# Domesticated

STAR --genomeDir data/salmonGenomeIndex --readFilesIn data/salmonTrimmedFastqsDomesticated/ERR3523019_trimmed.fastq.gz --outSAMtype BAM SortedByCoordinate --outFileNamePrefix data/salmonAlignedDomesticated/ERR3523019

mkdir data/salmonAlignedDomesticated
mkdir data/salmonAlignedWild

for i in {3522994..3522997} {3523017..3523018};
do STAR --genomeDir data/salmonGenomeIndex --readFilesIn data/salmonTrimmedFastqsDomesticated/ERR${i}_trimmed.fastq.gz --outSAMtype BAM SortedByCoordinate --outFileNamePrefix data/salmonAlignedDomesticated/ERR${i};
done

# Wild 
for i in {3523032..3523037};
do STAR --genomeDir data/salmonGenomeIndex --readFilesIn data/salmonTrimmedFastqsWild/ERR${i}_trimmed.fastq.gz --outSAMtype BAM SortedByCoordinate --outFileNamePrefix data/salmonAlignedWild/ERR${i};
done

```

Manual for star https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf
https://bioinformatics.uconn.edu/resources-and-events/tutorials-2/rna-seq-tutorial-with-reference-genome/

Method from https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4651-8#Sec2.

3. GENE COUNT
Generate raw gene counts using htsec.

```{bash}
# Domesticated

htseq-count -s no -r pos -t exon -i pacid -f bam data/salmonAlignedDomesticated/ERR3523019Aligned.sortedByCoord.out.bam data/salmonAnnotationGenome/GCF_000233375.1_ICSASG_v2_genomic.gff > data/salmonCountsDomesticated/ERR3523019-output_basename.counts


# Convert BAM Files to Raw Counts with HTSeq
for i in {3522994..3522997} {3523017..3523018};
do htseq-count -s no -r pos -t exon -i ID -f bam data/salmonAlignedDomesticated/ERR${i}Aligned.sortedByCoord.out.bam data/salmonAnnotationGenome/GCF_000233375.1_ICSASG_v2_genomic.gff > data/salmonCountsDomesticated/ERR${i}-output_basename.counts;
done

# Wild
# Convert BAM Files to Raw Counts with HTSeq
for i in {3523032..3523037};
do htseq-count -s no -r pos -t exon -i ID -f bam data/salmonAlignedWild/ERR${i}Aligned.sortedByCoord.out.bam data/salmonAnnotationGenome/GCF_000233375.1_ICSASG_v2_genomic.gff > data/salmonCountsWild/ERR${i}-output_basename.counts;
done

```


---------------------- R --------------------------------

4. NORMALIZE COUNT
CPM (Counts per million)
Filter the CPM counts so that we keep genes that have a CPM count >1 in more than 25% of the samples. 


For differential expression analysis between wild and domesticated, we use egdeR, using a full interaction model (see paper) -> output will be p-values 
Adjust p-values using FDR (?????), genes are deferentially expressed if  q < 0.05 and the absolute log2 fold change >1 between wild and domesticated (=> log2(wild_counts/domesticated_counts).


a. Do CPM
b. Rinse out genes with minimum count level of 1 CPM

```{r}
setwd("~/bioinfo/salmonproject/BBT045_AppliedBioinformatics")
if (!require("edgeR"))
    install.packages("edgeR")
library(edgeR)

# Domesticated
# character vector of filenames
# {3522994..3522997} {3523017..3523018}
fnames_domesticated <- c()
for (i in c(3522994:3522997, 3523017:3523018)) {
  fnames_domesticated <- append(
    fnames_domesticated, 
    paste("ERR", i, "-output_basename.counts", sep=""))
}

# Wild
# character vector of filenames
fnames_wild <- c()
# modify numbers
# {3523032..3523037}
for (i in c(3523032:3523037)) {
  fnames_wild <- append(
    fnames_wild, 
    paste("ERR", i, "-output_basename.counts", sep=""))
}

####### NEW ########

#We want to put the wild .counts files in the same directory as domesticated (salmonCountsDomesticated). Then put all the file names in the same vector so that when we run readDGE we get a DGE for ALL our files domesticated and wild! 

tot_fnames <- c(fnames_domesticated, fnames_wild)

group <- c("D","D","D","D","D","D","W","W","W","W","W","W") #domesticated =D, wild =W

rawcounts <- edgeR::readDGE(
  tot_fnames,  
  path="data/salmonCountsDomesticated",
  group = group,
  columns=c(1,2))

#Then we have to estimate the dispersion in order to fit a negative binomial model
dispersion <- edgeR::estimateCommonDisp(rawcounts) #output is new DGE list with new objects
names(dispersion)

#Exact test tests for negative binomial distribution by pairwise comparison between groups D and W
DE_genes <- edgeR::exactTest(dispersion)
names(DE_genes) 
names(DE_genes$tabel) #should give us "logConc", "logFC", "p.value"

ajd_pvalue <- edgeR::topTags(DE_genes, adjust.method = "BH", n=20, p.value = 0.05) #p.value is cut-off for adjusted p-values (q values), so from article it should be 0.05. Use sort.by = PValue (default)or logFC to sort the output. Adjusting through FDR See help page for more. Here we get the n=20 top differentially expressed genes

########### END NEW ##########

#If new code works, remove this!
# read file
#rawCountsDomesticated <- edgeR::readDGE(
#  fnames_domesticated, 
#  path="data/salmonCountsDomesticated",
#  columns=c(1,2))

# read file
#rawCountsWild <- edgeR::readDGE(
# fnames_wild, 
 # path="data/salmonCountsWild", 
 # columns=c(1,2))

#If new code works change line 302 & 303 to logCPM <- edgeR::cpm(rawcounts, log =TRUE), and change the PCA plots!

# raw count to CPM
logcpmCountsDomesticated <- edgeR::cpm(rawCountsDomesticated, log=TRUE) # prior.count=1
logcpmCountsWild <- edgeR::cpm(rawCountsWild, log=TRUE) # prior.count=1

# only CPM larger than 1
#rinsedLogcpmCounts <- logcpmCounts[rowSum(logcpmCounts > 1) <= ncol(logcpmCounts)*0.75, ]

# Differentially expressed genes
# ????
  
# PCA

pcaDomesticated <- prcomp(
  logcpmCountsDomesticated, 
  ntop = 2, 
  scale = TRUE, 
  verbose = TRUE)
summary(pcaDomesticated)
#https://rdrr.io/github/bedapub/ribiosNGS/man/prcomp.DGEList.html
# https://www.biostars.org/p/282685/

plot(
  pcaDomesticated$x,
  col = "orange")
plot(
  pcaWild$x,
  col = "skyblue",
  add=T)
legend(
  "bottomright",
  legend=c("Domesticated", "Wild"), 
  fill=c("red", "green"))

```


